[jayk.yang] 지하철 admin step4 리뷰 부탁드립니다.

tttuer opened the pull request:
domain에만 있던 클래스들을 각 패키지에 맞게 분류해보았습니다. 또한 controlleradvice를 이용하여 익셉션들을 관리할 수 있도록 하였습니다. 혹시 부족한 부분이 있으면 피드백 부탁드립니다. 감사합니다~!

Ye-Ji-Kim committed:
feat: CRUD 구현

Station CRUD, Line CRUD 구현

Ye-Ji-Kim committed:
feat: LineResponse default 생성자 추가

Ye-Ji-Kim committed:
feat: step2 지하철 구간 추가 API구현

tttuer committed:
feat: 1단계 실습

tttuer committed:
Merge branch 'step1' of https://github.com/Ye-Ji-Kim/spring-subway-admin-kakao into step1

Ye-Ji-Kim committed:
feat: 구간 등록 API 에러 수정

tttuer committed:
Merge branch 'step1' of https://github.com/Ye-Ji-Kim/spring-subway-admin-kakao into step1

Ye-Ji-Kim committed:
feat: step2 지하철 구간 추가 API 최종 수정

Ye-Ji-Kim committed:
feat: step3 jdbc로 변경

tttuer committed:
Merge branch 'step1' of https://github.com/Ye-Ji-Kim/spring-subway-admin-kakao into step1

tttuer committed:
feat: 버그 수정

Test실행 했을때 오류 수정

Ye-Ji-Kim committed:
feat: step3 구간 추가 API 에러 수정 및 구간 삭제 API 구현

tttuer committed:
feat: 안쓰는 것들 제거

tttuer committed:
Merge branch 'step1' of https://github.com/Ye-Ji-Kim/spring-subway-admin-kakao into step1

tttuer committed:
feat: 버그 수정

SectionService에서 삭제 시 오류 나던 부분 수정

tttuer committed:
refactor: 피드백 반영

도메인을 생성하여 서비스 부분을 축소, 라인 생성시 section도 같이 생성하도록 service에 트랜잭션 추가, exception 분류 및 null 리턴하지 않도록 수정, null을 확인하는 대신 optional을 이용하여 익셉션 발생하도록 수정, Response를 생성해주는 ResponseUtil 추가, 익셉션 변수 이름 수정, line 삭제시 그 안의 섹션들도 같이 삭제되도록 트랜잭션 추가

tttuer committed:
refactor: 피드백 반영

SectionRequest를 Section이 받아서 생성자에서 처리하도록 수정

tttuer committed:
refactor: 리팩토링

controller, dao, domain, dto, service로 리팩토링 진행, ControllerAdvice를 만들어서 Exception 핸들링

tttuer committed:
refactor: ControllerAdvice 이름 변경

tttuer committed:
refactor: customExceptionhandler 수정

각 exception별로 함수를 다르게 설정

tttuer committed:
refactor: CustomExceptionHandler 위치 이동

tttuer committed:
refactor: StationAlreadyExistException 추가

tttuer committed:
Merge branch 'tttuer' into step2

sehan commented on src/main/java/subway/line/service/LineService.java:
```diff
@@ -23,6 +26,9 @@ public LineService(LineDao lineDao, StationDao stationDao, SectionService sectio
+            throw new LineAlreadyExistException();
+        }
         Line newLine = lineDao.save(line);
         sectionService.insert(new Section(newLine.getId(), upStationId, downStationId, distance));
```
sectionService.insert() 내에 비지니스 로직이 남아 있기 때문에  SectionService 를 주입받아 사용하고 있는것으로 보입니다.
sectionService.insert() 내 비지니스 로직을 도메인으로 이동시켜주면 LineService 에서 SectionService 의존을 제거 하고 sectionDao 를 사용해서 LineService.insert() 코드를 작성가능합니다. 

아래에 있는 SectionService.showAll() 같은 경우에도 showAll() 메소드 내에 section 을 정렬하는 비지니스 로직이 있어서 SectionService 을 사용하고 있습니다. 비지니스 로직을 도메인쪽으로 이동시켜서 Service 에서 비지니스 로직을 제거 해주세요

sehan commented on src/main/java/subway/line/service/SectionService.java:
```diff
@@ -17,10 +19,8 @@ public SectionService(SectionDao sectionDao) {
+        return sections.sort(lineId, sections.getUpStationAndDownStation(), sections.getDownStationAndDistance());
     }
 
     public void insert(Section section) {
```
섹션 생성, 삭제 비지니스 로직이 service 에 많이 남아 있습니다.
도메인 객체로 비지니스 로직을 이동시켜 주세요.

도메인 객체로 이동된다면 아래 예시와 같은 형태로 SecionService.insert() 를 작성할 수 있습니다. 
```
    Line line = dao.findById(lineId)
    line.addSection( Station up, Station down, long distance )
    dao.save(line) 또는 sectionDao.saveAll(line.getSections())
```

sehan commented on src/main/java/subway/CustomExceptonHandler.java:
```diff
@@ -0,0 +1,41 @@
+import subway.station.domain.StationAlreadyExistException;
+
+@RestControllerAdvice
+public class CustomExceptonHandler {
```
controller 에서 try/catch 로 error 처리하던 코드들을 global exception handler 로 처리하니 아주 깔끔합니다. 👍 👍 👍

sehan reviewed (changes requested):
global exception handler 로 exception 처리를 잘 정리하셨습니다. 👍 👍 👍 

step4 는 layered architecture 구조로 리팩토링 하는것이 핵심입니다. 
아직 Service layer 에 비지니스 로직들이 많이 남아 있는 상태입니다. 
비지니스 로직들을 도메인쪽으로 이동하는 리팩토링을 더 진행해주세요 관련해서 코멘트 남겨놓았습니다. 

수정요청 드리겠습니다.

tttuer committed:
refactor: 피드백 반영

LineService에서 SectionServic를 가지지 않아도 되도록 Section domain에서 addSection 및 deleteSection을 하도록 수정

tttuer committed:
refactor: LineService에서 SectionService 삭제

sehan commented on src/main/java/subway/line/service/LineService.java:
```diff
@@ -12,19 +15,23 @@
+        }
         Line newLine = lineDao.save(line);
-        sectionService.insert(new Section(newLine.getId(), upStationId, downStationId, distance));
+        Sections sections = new Sections(sectionDao.showAllByLineId(newLine.getId()));
+        sections.addSection(new Section(newLine.getId(), upStationId, downStationId, distance));
+        sectionDao.deleteByLineId(newLine.getId());
+        sectionDao.saveAll(sections.getSections());
```
SectionService.insert 의 비지니스 로직을 도메인으로 잘 이동시켰네요. 아주 좋습니다. 👍

sehan commented on src/main/java/subway/line/service/LineService.java:
```diff
@@ -42,16 +49,18 @@ public void updateById(Long lineId, Line line) {
-        Sections sections = new Sections(sectionService.showAll(line.getId()));
-        List<Long> stationIds = sections.getStationIds();
-        stationIds.add(sections.getLastSectionDownStationId());
+        Sections sections = new Sections(sectionDao.showAllByLineId(line.getId()));
```
위에 LineService.insert() 에서 SectionService.insert() 의 비지니스 로직을 제거하고 객체간 협업으로 풀어낸 것처럼  노선의 역목록(정렬된) 을 제공해주는 책임을 LineService 가 가지는 것이 아니라 도메인객체에게 넘겨주면 비지니스 로직을 제거 할 수 있습니다.

sehan commented on src/main/java/subway/line/domain/SectionFactory.java:
```diff
@@ -0,0 +1,9 @@
+package subway.line.domain;
```
Controller 에서 RequestDTO -> Domain 객체로 변환해 주는 역할을 해주고 있습니다. 
domain 보다는 subway.line.dto 패키지에 속하는게 더 적절해 보입니다.

sehan reviewed (approved):
Service 에 있던 비지니스 로직을 도메인 으로 잘 이동시켜주셨습니다. 👍 👍 👍 

끝까지 비지니스 로직 분리에 힘써 주셔서 감사합니다. 지하철 admin 미션은 여기서 merge 하도록 하겠습니다. 이후 지하철 서비스 미션도 힘내서 진행 잘 해주시기 바랍니다. 

미션종료와 관계 없이 몇가지 가볍게 살펴볼 만한 코멘트 남겨두었습니다.  고생하셨습니다. ☺️
